src/pages/api/webhooks/resource-training.ts

```diff
@@ -0,0 +1,208 @@
+import { TrainingStatus } from '@prisma/client';
+import * as z from 'zod';
+import { env } from '~/env/server.mjs';
+import { dbWrite } from '~/server/db/client';
+import { trainingCompleteEmail } from '~/server/email/templates';
+import { refundTransaction } from '~/server/services/buzz.service';
+import { WebhookEndpoint } from '~/server/utils/endpoint-helpers';
+
+export type EpochSchema = z.infer<typeof epochSchema>;
+const epochSchema = z.object({
+  epoch_number: z.number(),
+  model_url: z.string(),
+  sample_images: z
+    .array(
+      z.object({
+        image_url: z.string(),
+        prompt: z.string(),
+      })
+    )
+    .optional(),
+});
+
+type ContextProps = z.infer<typeof context>;
+const context = z.object({
+  modelFileId: z.number().gt(0),
+  status: z.string(),
+  message: z.string().optional(),
+  model: z.string(),
+  start_time: z.number(),
+  end_time: z.number().optional(),
+  duration: z.number().optional(),
+  epochs: z.array(epochSchema).optional(),
+  upload_duration: z.number().optional(),
+  sample_prompts: z.array(z.string()).optional(),
+  logs: z
+    .object({
+      stdout: z.string().optional(),
+      stderr: z.string().optional(),
+    })
+    .optional(),
+});
+
+const schema = z.object({
+  jobId: z.string(),
+  type: z.string(),
+  dateTime: z.string(),
+  // serviceProvider: z.string().nullish(),
+  workerId: z.string().nullish(),
+  context: context.nullable(),
+  claimDuration: z.string().nullish(),
+  jobDuration: z.string().nullish(),
+  retryAttempt: z.number().nullish(),
+  // cost: z.number().nullish(),
+  jobProperties: z.object({
+    transactionId: z.string(),
+    // userId: z.number(),
+  }),
+  jobHasCompleted: z.boolean(),
+});
+
+// breaking change
+export default WebhookEndpoint(async (req, res) => {
+  if (req.method !== 'POST') {
+    return res.status(405).json({ error: 'Method not allowed' });
+  }
+
+  const bodyResults = schema.safeParse(req.body);
+  if (!bodyResults.success) {
+    return res.status(400).json({ ok: false, errors: bodyResults.error });
+  }
+
+  const data = bodyResults.data;
```

Comment:
I think it'd be nice to get this typed with ZOD - Not a stopper here or anything, but pointing it out in case you didn't consider it

Reply:
it is typed with zod in the above schema